### PR TITLE
Key-region lock manager

### PR DESCRIPTION
**Pull request by johannkm**
The large diff is the deprecated lock manager.

- proto_utils: remaster transactions are marked as MH
- sequencer: at the replica of new_master, create a new lock only
- scheduler: only changes are to not use the remaster manager if counterless remastering
- lock manager: request locks on <key, replica> pairs. Remaster txns need special behavior: the new lock only requests for <key, new master> and releasing the txn releases both locks
- worker: check masters

This leaves some potential optimizations for the future. Transactions can always acquire locks, even when they will be aborted later. Locks for old masters aren't deleted automatically

Testing:
- basic 2 test cases in deterministic_lock_manager_test. Also removed 2 cases which had 'ghost transactions' with no keys
- scheduler
- e2e
- running the remaster benchmark completes, doesn't necessarily imply correctness

### CODE REVIEW

File: module/scheduler_components/deterministic_lock_manager.h [hunk 1]
```diff
@@ -20,12 +28,17 @@ using std::vector;
 
 namespace slog {
 
+using KeyReplica = string;
+
+enum class AcquireLocksResult {ACQUIRED, WAITING, ABORT};
```
Comment (johannkm) [inline module/scheduler_components/deterministic_lock_manager.h:33]: This implementation doesn't use aborts here, but I think we could leave it in for future optimizations

File: module/sequencer.cpp [hunk 2]
```diff
@@ -162,6 +163,23 @@ void Sequencer::ProcessMultiHomeBatch(Request&& req) {
...
+    lock_only_txn->mutable_remaster()->set_new_master((txn.remaster().new_master()));
+    // Add additional lock only at new replica
+    // TODO: refactor to remote metadata from lock-onlys. Requires
+    // changes in the scheduler
+    if (txn.procedure_case() == Transaction::kRemaster) {
+      if (txn.remaster().new_master() == local_rep) {
+        auto key_value = *txn.write_set().begin();
+        lock_only_txn->mutable_write_set()->insert(key_value);
+        auto& new_metadata = (*lock_only_metadata)[key_value.first];
+        new_metadata.set_master(metadata.at(key_value.first).master());
```
Comment (ctring) [inline module/sequencer.cpp:177]:
```C++
lock_only_txn->CopyFrom(txn)
```
Then move the lines for `set_id` and `set_type` above to down below this section

File: module/sequencer.cpp [hunk 3]
```diff
@@ -162,6 +163,23 @@ void Sequencer::ProcessMultiHomeBatch(Request&& req) {
         lock_only_metadata->insert({key_value.first, metadata.at(key_value.first)});
       }
     }
+
+#ifdef REMASTER_PROTOCOL_COUNTERLESS
+    lock_only_txn->mutable_remaster()->set_new_master((txn.remaster().new_master()));
```
[the remark this answers is not in the document]
Comment (johannkm) [reply module/sequencer.cpp]: Good catch, i'm surprised this didn't break everything. I guess if code is already set then it won't be overwritten?

[the remark this answers is not in the document]
Comment (johannkm) [reply module/sequencer.cpp]: Documentation says this should have been overriding code, not sure how benchmarks were able to complete. Maybe I had 0% mh.

**Pull request merged**